core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1531,6 +1531,8 @@ class Partition(val topicPartition: TopicPartition,
             isrState = CommittedIsr(leaderAndIsr.isr.toSet)
             zkVersion = leaderAndIsr.zkVersion
             info(s"ISR updated to ${isrState.isr.mkString(",")} and version updated to [$zkVersion]")
+            leaderLogIfLocal.exists(leaderLog => maybeIncrementLeaderHW(leaderLog))
```

Comment:
Should we only do this upon it's a shrink request (i.e. `proposedIsrState` is a `PendingShrinkIsr`)?

Underlyingly maybeIncrementHighwatermark has a `lock.synchronized` block; this should reduce lock contention